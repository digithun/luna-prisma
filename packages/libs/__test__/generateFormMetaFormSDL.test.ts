import {
  parseSDLToFormMeta,
  findTypeDefinitionInfoFromSchema,
} from "../generateFormMetaFromSDL"
import gql from "graphql-tag"
import { buildClientSchema, printSchema, parse } from "graphql"

const schema = require("./luna-schema")

it("should create FormMeta correctly", () => {
  expect(
    parseSDLToFormMeta(
      gql`
        query($where: TodoWhereUniqueInput!) {
          todo(where: $where) @form {
            id @input(label: "ID")
            name @input
            color @input(label: "Colour")
          }
          t: todo(where: $where) {
            id
            name
          }
        }
      `,
      schema
    )
  ).toMatchSnapshot()
})

it("should findTypeDefinitionFromPath correctly", () => {
  const schemaAST = parse(printSchema(buildClientSchema(schema)))
  expect(
    findTypeDefinitionInfoFromSchema(schemaAST, [
      {
        kind: "OperationDefinition",
        operation: "query",
      },
      {
        kind: "Field",
        name: { value: "todo" },
      },
      {
        kind: "Field",
        name: { value: "name" },
      },
    ] as any)
  ).toEqual({
    isEditable: true,
    isNonNull: true,
    kind: "FormTextInputFieldMeta",
    path: "query.todo.name",
    key: "name",
    typeName: "String",
  })

  expect(
    findTypeDefinitionInfoFromSchema(schemaAST, [
      {
        kind: "OperationDefinition",
        operation: "query",
      },
      {
        kind: "Field",
        name: { value: "todo" },
      },
      {
        kind: "Field",
        name: { value: "color" },
      },
    ] as any)
  ).toEqual({
    isEditable: true,
    isNonNull: false,
    kind: "FormEnumInputFieldMeta",
    key: "color",
    path: "query.todo.color",
    options: ["RED", "BLUE", "GREEN"],
    typeName: "ColorEnum",
  })

  expect(
    findTypeDefinitionInfoFromSchema(schemaAST, [
      {
        kind: "OperationDefinition",
        operation: "query",
      },
      {
        kind: "Field",
        name: { value: "todo" },
      },
      {
        kind: "Field",
        name: { value: "color" },
      },
    ] as any)
  ).toEqual({
    isEditable: true,
    isNonNull: false,
    kind: "FormEnumInputFieldMeta",
    key: "color",
    path: "query.todo.color",
    options: ["RED", "BLUE", "GREEN"],
    typeName: "ColorEnum",
  })
})
