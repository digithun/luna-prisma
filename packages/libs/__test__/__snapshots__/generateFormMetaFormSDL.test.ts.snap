// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create FormMeta correctly 1`] = `
Array [
  Object {
    "isEditable": false,
    "isNonNull": true,
    "key": "id",
    "kind": "FormTextInputFieldMeta",
    "path": "query.todo.id",
    "typeName": "ID",
  },
  Object {
    "isEditable": true,
    "isNonNull": true,
    "key": "name",
    "kind": "FormTextInputFieldMeta",
    "path": "query.todo.name",
    "typeName": "String",
  },
  Object {
    "isEditable": true,
    "isNonNull": false,
    "key": "color",
    "kind": "FormEnumInputFieldMeta",
    "options": Array [
      "RED",
      "BLUE",
      "GREEN",
    ],
    "path": "query.todo.color",
    "typeName": "ColorEnum",
  },
  Object {
    "isEditable": true,
    "isNonNull": false,
    "key": "description",
    "kind": "FormTextInputFieldMeta",
    "path": "query.todo.description",
    "typeName": "String",
  },
]
`;
